### ISSUE TITLE
latent token and CoT match = 0

### ISSUE THREAD

**Issue by Sumireko-Usami** [facebookresearch/coconut#31]
![Image](https://github.com/user-attachments/assets/50683419-2d3e-4768-93db-a0a4e1bc84b9)

During the training and validation process on GSM8K, I noticed that in the validation phase, the model's full output still contains the latent tokens, instead of the actual visible CoT. Is this expected behavior? Also, the "CoT match on validation set" metric always remains at 0 — is that normal as well?

**Comment by ExtraBits42** [comment]
Hello, I suggest you take a look at the implementation of the source code. In fact, supervised learning is used for training here, and identifiers such as<| latent |>are actually inserted into the prompt for data preprocessing. During the model inference process, internal latent space inference is performed based on these special identifiers in the prompt. Also, please clarify whether the GSM8K training and inference you are using is the CoT version, Coconut version, or any other version.
By the way, in fact, after conducting a series of experiments, I have some doubts about the effectiveness and principle explanation of this work. I think the so-called continuous latent space reasoning may not necessarily be truly useful for improving the effectiveness of LLM.

**Comment by Ber666** [comment]
> ![Image](https://github.com/user-attachments/assets/50683419-2d3e-4768-93db-a0a4e1bc84b9)
> 
> During the training and validation process on GSM8K, I noticed that in the validation phase, the model's full output still contains the latent tokens, instead of the actual visible CoT. Is this expected behavior? Also, the "CoT match on validation set" metric always remains at 0 — is that normal as well?

Yes, the model is expected to use continuous thought to conduct reasoning. "CoT match is only an auxiliary metric to understand the SFT training, and is expected to be zero after coconut training, because the model will use latent space instead of language cot.

**Comment by Ber666** [comment]
> Hello, I suggest you take a look at the implementation of the source code. In fact, supervised learning is used for training here, and identifiers such as<| latent |>are actually inserted into the prompt for data preprocessing. During the model inference process, internal latent space inference is performed based on these special identifiers in the prompt. Also, please clarify whether the GSM8K training and inference you are using is the CoT version, Coconut version, or any other version.
> By the way, in fact, after conducting a series of experiments, I have some doubts about the effectiveness and principle explanation of this work. I think the so-called continuous latent space reasoning may not necessarily be truly useful for improving the effectiveness of LLM.

I don't get your point. Please let me know if there are any specific questions about the paper or the code.
